lib/delayed/backend/active_record.rb

```diff
@@ -109,11 +109,18 @@ def self.reserve_with_scope_using_optimized_sql(ready_scope, worker, now)
         end
 
         def self.reserve_with_scope_using_default_sql(ready_scope, worker, now)
-          # This is our old fashion, tried and true, but slower lookup
-          ready_scope.limit(worker.read_ahead).detect do |job|
-            count = ready_scope.where(id: job.id).update_all(locked_at: now, locked_by: worker.name)
-            count == 1 && job.reload
+          # This is our old fashion, tried and true, but possibly slower lookup
+          # Instead of reading the entire job record for our detect loop, just pluck the ID,
+          # and only read the job record after we've successfully locked the job.
+          # This can be particularly helpful when operating a large job cluster that uses
+          # a large read_ahead value to increase the odds of successfully locking a job with
+          # one method call to reserve_with_scope.
+          locked_job_id = ready_scope.limit(worker.read_ahead).pluck(:id).detect do |job_id|
```

Comment:
`pluck` was added in Rails 3.2 and `find_by` was added in Rails 4.0. We still have compatibility back to Rails 3.0 and this doesn't feel like a good enough reason to break that. By only adding `select(:id)` to the original we can get almost all the benefits of your change without dropping backward compatibility. `pluck` is slightly more object efficient, but probably not enough to matter here.

Nice catch on loading a lot more than we need here.

Reply:
Good callout, thanks. I've pushed an update that will hopefully resolve the compatibility issue.

Reply:
You should be able to go back to the original `reload` version vs the `find`. The `reload` should populate the entire object

Reply:
Right you are.